server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -128,6 +128,15 @@ public String name() {
         return name;
     }
 
+    /**
+     * Returns the lucene field name, defaults to name() needs to be overwritten
+     * if mapping field name is different from lucene field name.
+     * @return the lucene field name
+     * */
+    public String concreteFieldName() {
+        return name();
+    }
```

Comment:
While all of the name vs concrete field name usages in the subclasses don't matter that much (because in practice they are most of the times the same thing), in the base class, I think we should review all the usages of `name()`. They all look correct in that they keep on using `name()` which I view as the user defined name, while the concrete one is the internal name that should not leak out to users.

The `existsQuery` method though, when it performs a term query on _field_names, uses `name()`: that relies on the fact that `name()` is also used when storing the field names in it, which I see in some places is switched to concrete field name. Also, the definition of this method expects that existsQuery for fields that are special is overridden, like it is the case in rank feature field type, so it is not in practice a problem. What gets me anxious is that we don't have many guarantees because in many places name == concrete.